### PR TITLE
Articles can be downloaded and resumed

### PR DESCRIPTION
**Pull request by todorus**
PR for #16  

* Added the SqlAlchemy model that is used in the hunter-gatherer repo, so we both use the same model (I need to move this to a seperate repo, so it can be a seperate pip library instead)
* Articles can be streamed to a csv file by using data.collect.download_articles_to_file
* Downloads can be resumed when they are interrupted
* The word2vec model generation can load the data from a csv file in a streaming manner

**example of how to download articles**
```
from data import model
from data import collect

engine = model.engine(
    database_user=database_user,
    database_password=database_password,
    database_host=database_host,
    database_port=database_port,
    database_name=database_name
)

session = model.session(engine)

collect.download_articles_to_file(session, "articles.csv")
```

Then afterwards you can run ```python createModel.py input.csv output.w2vec``` to generate the model

### CODE REVIEW

**Comment by Tilana** [comment]
Great! Some things: MAybe we can find a better name for the data module. Data is too generic and often used as a folder to save raw data files...

File: newsAnalysis/data/collect.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                last_id = int(row[0])
+    print ("found latest id %d" % last_id)
+
+    query = session.query(Article)\
+                .filter(Article.body.isnot(None))\
+                .filter(Article.id >= last_id)\
+                .filter(Article.source_id == source_id)\
+                .order_by(Article.id)\
+                .yield_per(10)
+
```
Comment (Tilana) [inline newsAnalysis/data/collect.py]: Can we add here a parameter to indicate the number of documents that we query?
This is important to test functions with very small datasets and also to keep the number of documents from different newspapers constant

File: newsAnalysis/data/collect.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+from data.model import Article
+import unicodecsv as csv
+import os
+
+
+def download_articles_to_file(session, file_path, source_id=1):
+    '''
+    connection
+        A psycopg2 connection
+    file_path : String
```
Comment (Tilana) [inline newsAnalysis/data/collect.py]: where is the psycopg2 connection? Do we still need that module?

File: requirements.txt [hunk 3]
```diff
@@ -5,6 +5,8 @@ Pattern
 wordcloud
 pygal
 nose
-psycopg2
-
-
+psycopg2==2.7.1
```
Comment (Tilana) [inline requirements.txt:8]: remove?

File: newsAnalysis/data/collect.py (hunk 2 above)
Comment (todorus) [reply newsAnalysis/data/collect.py]: I need to update that documentation to say that is a SqlAlchemy connection

File: requirements.txt (hunk 3 above)
Comment (todorus) [reply requirements.txt:8]: It is needed. It is used by sqlalchemy to support features specific to postgresql. It's called a 'dialect' of sql in sqlalchemy terms.

**Comment by todorus** [comment]
Instead of a 'data' folder, what about 'datastore'?

File: requirements.txt (hunk 3 above)
Comment (Tilana) [reply requirements.txt:8]: ok. perfect

**Comment by Tilana** [comment]
datastore is perfect!

File: requirements.txt (hunk 3 above)
Comment (Tilana) [reply requirements.txt:8]: ok.. got it

File: newsAnalysis/datastore/collect.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    print ("%d aricles have already been downloaded" % downloaded_articles_count)
+    print ("found latest id %d" % last_id)
+
+    corrected_limit = None
+    if limit is not None:
+        corrected_limit = max(0, limit - downloaded_articles_count)
+    print("corrected_limit = %r" % corrected_limit)
+
+    if corrected_limit is None or corrected_limit > 0:
+        query = session.query(Article)\
```
Comment (Tilana) [inline newsAnalysis/datastore/collect.py:43]: Limiting the number of documents and adding new ones to the csv file work like a charm!
However, I get some, ~10%, empty rows (only id and source_id, but no published_date, title or body) in the csv.
Maybe an issue with query and its filter?

File: newsAnalysis/datastore/collect.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                    .filter(Article.source_id == source_id)\
+                    .order_by(Article.id)\
+                    .limit(corrected_limit)\
+                    .yield_per(10)
+
+        with open(file_path, "ab") as f:
+            writer = csv.writer(f, encoding='utf-8')
+            if not file_exists:
+                writer.writerow(['id', 'source_id', 'published', 'title', 'body'])
+            for article in query:
```
Comment (Tilana) [inline newsAnalysis/datastore/collect.py:55]: I added the column names to the csv file for an easier access later on

File: newsAnalysis/datastore/collect.py (hunk 4 above)
Comment (todorus) [reply newsAnalysis/datastore/collect.py:43]: Do you have any example ids?

Comment (todorus) [reply newsAnalysis/datastore/collect.py:43]: I don't think it's the filter as less then 1% of rows has incomplete data. That would be a pretty big coincidence. Perhaps something goes wrong when writing or reading (special chars maybe).

I will write a test for those ids and look into it. If csv is unreliable I'll switch to binary

Comment (Tilana) [reply newsAnalysis/datastore/collect.py:43]: I get 9 empty rows out of 100. Sample ids are 3,9,16,22

Comment (Tilana) [reply newsAnalysis/datastore/collect.py:43]: Maybe an encoding issue with the csv writer? But then it's strange that there is no date shown...

**Pull request merged**